api/server/router/container/container_routes.go

```diff
@@ -254,18 +253,8 @@ func (s *containerRouter) postContainersKill(ctx context.Context, w http.Respons
 		return err
 	}
 
-	var sig syscall.Signal
 	name := vars["name"]
-
-	// If we have a signal, look at it. Otherwise, do nothing
-	if sigStr := r.Form.Get("signal"); sigStr != "" {
-		var err error
-		if sig, err = signal.ParseSignal(sigStr); err != nil {
-			return errdefs.InvalidParameter(err)
-		}
-	}
-
-	if err := s.backend.ContainerKill(name, uint64(sig)); err != nil {
+	if err := s.backend.ContainerKill(name, r.Form.Get("signal")); err != nil {
```

Comment:
Using a string as "hand-over" variable between the API and the backend; this keeps the `syscall` package in the backend (which felt more appropriate than inside the API, also because it already needed it there)

Reply:
^^ FWIW I'm a bit on the fence on this specific change;

- 👍 keeps syscall out of the API bit (may not make a lot of sense to have it here)
- 👎 "backend" interface is a bit less "strict" (no real strong type); although the existing `int` wasn't great either (perhaps changing this to a `syscall.Signal` as well??)
 
So, happy to get feedback on this one

Reply:
^^ FWIW I'm a bit on the fence on this specific change;

- 👍 keeps syscall out of the API bit (may not make a lot of sense to have it here)
- 👎 "backend" interface is a bit less "strict" (no real strong type); although the existing `int` wasn't great either (perhaps changing this to a `syscall.Signal` as well??)
 
So, happy to get feedback on this one

Reply:
I am completely in favor of using a string as the hand-over between the API and backend as it allows the backend to parse the signal in a platform-specific manner. For example, SIGRTMIN can be different on different platforms so by passing string-typed signals an e.g. hypothetical future Darwin backend could parse `"SIGRTMIN+3"` to a different signal number than a Linux backend. Pushing parsing into the backend would also allow multiple backends with different signal-parsing semantics to coexist in the same daemon, so e.g. a hypothetical future Docker Desktop for Mac could support both Darwin and Linux(kit) containers concurrently.

Reply:
Thanks!

So, I later realised that I already took the same approach for `backend.ContainerStop()`; https://github.com/moby/moby/blob/41b96bff556e1467882f98a657b8396afaee6733/api/server/router/container/backend.go#L43

Which takes a StopOptions, which uses a string for the signal as well; https://github.com/moby/moby/blob/41b96bff556e1467882f98a657b8396afaee6733/api/types/container/config.go#L16-L21


> For example, SIGRTMIN can be different on different platforms so by passing string-typed signals an e.g. hypothetical future Darwin backend could parse "SIGRTMIN+3" to a different signal number than a Linux backend.

I definitely agree on using a string for this when crossing platform boundaries, and proposed a similar approach in containerd https://github.com/containerd/containerd/issues/5931#issuecomment-912006262 / https://github.com/containerd/go-runc/pull/78)

In this case (currently) there was no crossing platforms, which put me a bit on the wrong foot wondering "is this the right thing"

Given that we already take this approach, and the other (potential future) advantages, I think this is good to go 👍